### DISCUSSION TITLE
Possibility to design "Note to self"

### DISCUSSION THREAD

**Discussion by mnalis** [streetcomplete/StreetComplete#2559] [Ideas]
Let's try to constructively design how we would want it to work, originally from https://github.com/streetcomplete/StreetComplete/issues/1992

Feel free to like/dislike each of the ideas for each of the tasks needed to solve the issue, or propose your own as a reply to the problem (I've tried to add main ones from #1992 but I've probably missed some)!

Main things to think about:
- more user friendly is better (SC is built around user-friendliness as a main principle!)
- more complete / more functional feature is better, unless it significantly detracts from user-friendliness
- if ideas are comparable, the one which is more simple to implement is better

**Comment by mnalis** [comment]
## How to create a "private note"? 
... and things to take in account while doing that.

**Reply by mnalis** [reply]
- Main idea seems to be to use the same path as for "public note" (eg. "long press on screen" or answering "Other/Can't say" on quest), and then have a checkbox marked "private". Mockup screenshot would be nice here.

**Reply by mnalis** [reply]
- private notes should be flagged as such, so they're never uploaded, and pictures should be in full size and quality (instead of [reduced size](https://github.com/streetcomplete/StreetComplete/issues/2497#issuecomment-760214653) and [quality](https://github.com/streetcomplete/StreetComplete/issues/2538) which public notes undergo to be published on the net)

**Reply by rhhsm** [reply]
> 
> 
>   pictures should be in full size and quality (instead of reduced size and quality which public notes undergo to be published on the net)

Doesn't matter, reduced size is still more than "good enough" to use, but if it's less trouble not to reduce them, that's fine too.

**Reply by mnalis** [reply]
[v32](https://github.com/streetcomplete/StreetComplete/releases) (in disabled-auto-upload mode) allows users to create notes, and then selectively delete them before uploading. In that case, private note is created the same as normal public note, but user deletes it before manually initiating upload.

**Reply by mnalis** [reply]
- one can use `long press` + `Open location in another app` in order to open note in some other app like [notepad which supports `geo:` functionality](https://github.com/streetcomplete/StreetComplete/issues/1992#issuecomment-839141417). In such case, no additional support in SC is needed.

**Reply by Helium314** [reply]
Commenting here, even though it applies to more parts of this discussion:
As you noted, using v32 could remove the need for private notes as they can be checked before uploading.

So there could be a setting like _don't upload new notes immediately_, and when active one needs to tap the upload button for new notes to be uploaded. This makes finding notes in edit history much easier.
If auto-upload is disabled, upload button only upload notes if there is no quest/note comment to upload... maybe some indicator separating number of quests and number of notes might be useful here.
Or maybe the option could stop upload of new notes at all to avoid accidental upload. Or show a dialog...

Edit history would need to be changed a bit for nicer display of notes + images.

So now the notes can be checked and removed manually, or simply uploaded.
An export button/option could be implemented, showing the folder where notes should be stored as gpx + jpg.

**Comment by mnalis** [comment]
## How to promote private note to public note?
(when promoted to public notes, pictures are resized and uploaded immediately if autoupload is enabled)

**Reply by mnalis** [reply]
- one of the ideas is to select a note (in the same way as if you wanted to comment on it) and have an extra options there. How exactly would that work and look (screenshot mockup) ?

**Reply by rhhsm** [reply]
Maybe the checkbox that marks it as a private note can be unchecked to upload it as public note? Low priority additional feature

**Reply by mnalis** [reply]
[v32](https://github.com/streetcomplete/StreetComplete/releases) (in disabled-auto-upload mode) allows users to create notes, and then selectively delete them before uploading. In that case, private note is created the same as normal public note, but users are expected to delete private notes before manually initiating upload - if they do not, the note is "promoted" to normal public note.

**Comment by mnalis** [comment]
## How to find private notes created?

**Reply by mnalis** [reply]
- no functionality is needed, user needs to be able to find his own quest without any help -- just let him/her use regular quest view (would probably be too unfriendly to the user, as other quests would make it hard to find a tree in a forest) 

As example how regular quest view is unsuitable for the purpose, try to find notes in this picture:
![notes1](https://user-images.githubusercontent.com/156656/107162813-0be55a80-69a6-11eb-8c87-ee86deb5d824.png)

- This idea also has a problem that some notes created by user are hidden from him on purpose in regular quest view. If implemented this way, the reasons for that and workarounds for that would first need to be found.

- if still interested in this way, how should a private notes be shown? same icon/color as public notes or different? (Screenshot mockups please).

**Reply by mnalis** [reply]
- have a special view on the map which shows only private notes? How would it be invoked (mockup needed)

**Reply by mnalis** [reply]
- have a list which zooms in on the map on that note when clicked (something like [https://my-notes.osm-hr.org/](https://my-notes.osm-hr.org/)). Advantage - that list can also be used for deleting notes by having `delete` button in addition to `zoomin/show` button.

[v32](https://github.com/streetcomplete/StreetComplete/releases) (in disabled-auto-upload mode) allows users to create notes, and then selectively delete them before uploading. This effectively implements this option. User clicks on `undo button` and the selects `Show edit history`, where all notes and quests are displayed, and user can click on each one to center map display on it and show note content, to manually add it in desktop editor before deleting it.

**Reply by rhhsm** [reply]
Public notes now have an extra high priority on SC so they are always shown on top of any other quest, isn't it? Private notes could be given the same or next-highest priority.

**Reply by mnalis** [reply]
@rhhsm I've also read in some ticket that Notes are given higher tangrem-es priority, but that feature does not seem to work reliably to me - either it is just "best effort", or it is buggy. Here is example of same area just slightly less zoomed - sometimes Notes are shown, sometimes not. I do not think that is something that can be counted on at the moment...

![a](https://user-images.githubusercontent.com/156656/107306092-85537a80-6a84-11eb-952f-6a735991df41.png) ![b](https://user-images.githubusercontent.com/156656/107306086-8389b700-6a84-11eb-9434-f309a9b69415.png)

**Reply by rhhsm** [reply]
I expect that it won't be a problem in practice. Notes will be made during a survey, and many quests will be solved during that same survey so that the map will be relatively empty of quests right where the notes are. I also don't see it as a problem if occasionally a mapper misses a note or two because he can't find them. They will show up sooner or later.

**Reply by mnalis** [reply]
I'm not so sure it isn't a problem. See usage cases for quests presets for example (https://github.com/streetcomplete/StreetComplete/discussions/2457), only a very few of them involve solving all quests in the area, in most cases only some (even very few) of the quests types are being solved. Also, such prediction depends much on mapper density - for exapmle, I do not expect quest map to EVER clear in my area, as before most of the quests are solved by existing mappers, they'd pop up again as resurvey quests. And that is not even counting new versions of SC which constantly add new quests.

**Reply by Strubbl** [reply]
If private notes are a type of quest like the public notes, one could put it at the beginning of the quest priority list in the settings "task selection". Icon can be the same like the public note but maybe a different color than red, e.g. blue instead of red.

**Comment by mnalis** [comment]
## How to delete private notes when finished with them?
Needed so user knows what has s/he solved, and to free space taken by (full-size and full-quality) photos

**Reply by mnalis** [reply]
- if using list view of all private notes like https://my-notes.osm-hr.org/, there could be a separate `delete` button

[v32](https://github.com/streetcomplete/StreetComplete/releases) (in disabled-auto-upload mode) allows users to create notes, and then selectively delete them before uploading. This effectively implements this option. User clicks on `undo button` and the selects `Show edit history`, where all notes and quests are displayed, and user can click on each one to center map display on it and show note content, to manually add it in desktop editor before deleting it in SC. When all notes deemed private are deleted, user can do upload, and all remaining notes (the ones deemed public) will get uploaded with quests.

**Reply by mnalis** [reply]
- if notes could be "edited" by clicking on them (ie. there would be `promote to public note` and `delete` buttons, and not just `comment` button it has now). (Mockup screenshots needed)

**Reply by mnalis** [reply]
- if using "share" to desktop computer, the user could be asked after transferring all the files to desktop computer if s/he wants to delete them from mobile device

**Reply by rhhsm** [reply]
By tapping a public note on SC, there is the option to hide it. Similarly, when tapping a private note there could be the option to delete it (at the same spot where the "No, hide" option is for public notes).

**Comment by mnalis** [comment]
## How to use private notes to actually solve them?

**Reply by mnalis** [reply]
- no implementation needed, user will just use JOSM/iD on their computer, and look at the mobile phone at the side, manually finding notes and manually finding same location in desktop editor (problems: error prone, time consuming). This would also need some kind of picture viewer/zoomer on mobile phone (where exactly would it be put? mockup screenshots please)
- this solution also have the disadvantage of using SMALL mobile screen instead of BIG desktop screen to look at the pictures, thus making it harder to make out details etc (text on photographs, etc)


This option is effectively implemented in [v32](https://github.com/streetcomplete/StreetComplete/releases) (in disabled-auto-upload mode). User clicks on `undo button` and the selects `Show edit history`, where all notes and quests are displayed, and user can click on each one to center map display on it and show note content, to manually add it in their prefered desktop editor before deleting it.

**Reply by mnalis** [reply]
- use android "share" functionality (which supports many ways of file transfer) to send .GPX file (with position and text of notes) and geolocated .JPG files (pictures) so user can open then in desktop editor and there are positioned at right places on the map. (advantages: no errors in location; it might also solve deleting task by asking after transfer if data should be removed from mobile)

**Reply by rhhsm** [reply]
Hmmm, that's indeed a tricky one I hadn't thought of. The ideal solution would be if private notes also appear in the editor on the big screen, but only if the author of the note is logged in in that editor. That would need uploading by SC and design development in those editors. But maybe users of JOSM and iD would also appreciate to be able to make notes on the map for themselves only?

**Reply by FloEdelmann** [reply]
Is it feasible to reach a change to the OSM (Notes) API? Then "private" could be a property of the note, nothing would need to be changed in other editors (they'd still see all notes), and the upload logic in StreetComplete could stay the same.

Or maybe a consensus could be reached on treating all notes starting with `@osmUserName` as private.

**Reply by mnalis** [reply]
@FloEdelmann While [API](https://wiki.openstreetmap.org/wiki/API_v0.6) change is possible (it had about 7 changes in last 12 years), it wouldn't mean anything without ALSO at least biggest of few dozen most popular [editors](https://wiki.openstreetmap.org/wiki/Editors) adopting it, for which I guess idea should be much more widely popular (eg. several hundred users demanding it of their favorite editors, who would then ask it of OSMF to implement)... 

@rhhsm if you only share data locally from your mobile to your desktop (using [Android share button](https://commons.wikimedia.org/wiki/File:Font_Awesome_5_solid_share-alt-square.svg), for example via e-mail or bluetooth or ShareViaHTTP or whatever) then you don't need editors to be modified to take advantage of received data (at least JOSM doesn't need to be changed; I know iD allows GPX with named points to be imported which can be used for text notes, but I'm not sure if supports showing pictures automatically as JOSM does).  

If you instead upload your notes somewhere outside of your own computer, then they're [not really "private"](https://planet.openstreetmap.org/notes/), even if they might be "not showing to other users by default" (and even that is assuming enough users can persuade various OSM Editors and OSMF servers to change the API)

**Reply by Strubbl** [reply]
It would be nice if i can export all private notes incl. their location, text and attached pictures. [OsmTracker](https://github.com/labexp/osmtracker-android) has a similar solution for export. This way i can download the export to my PC and use JOSM to edit.

After export and download to my PC, i want to go to the SC app and delete all private notes at once.

**Reply by mnalis** [reply]
> Or maybe a consensus could be reached on treating all notes starting with @osmUserName as private.

according to [this post](https://community.openstreetmap.org/t/tags-denoting-note-status-and-note-comments/4788), one could add `#notetoself` in the note text to indicate so.

**Comment by mnalis** [comment]
## Misc. discussion

If you don't find a better place, feel free to reply here to ask or express opinions etc.

**Reply by Helium314** [reply]
I recently added a VERY basic version of private notes (using a gpx file) into my modified version of SC, the note part is available here: https://github.com/Helium314/StreetComplete/tree/gpx_notes

This allows choosing between OSM notes and GPX when creating a note. OSM will behave like usual; GPX will not remove the involved quest (if any) and add the note to the file notes.gpx in \<internal storage\>/Android/data/de.westnordost.streetcomplete/files
The file is immediately written when creating a note, without any further checks. So there is a chance of crashing if the file can't be written (never happened to me).

Images are completely ignored, better don't try using them when creating a GPX note (I didn't check and they might just be stored inside some SC folder for all eternity).

This is not really meant to be included in SC, more for being able to do some "live testing" of private notes.

**Reply by mnalis** [reply]
@Helium314 it would be nice if you could make `APK` on Release page, so people without full android development environment might try it out and give comments.

**Reply by Helium314** [reply]
see https://github.com/Helium314/StreetComplete/releases/tag/v31.1_notes

**Comment by mnalis** [comment]
## Why is it needed?

**Reply by mnalis** [reply]
- other mappers complain that you shouldn't be leaving public notes if you plan to solve them yourself (as it is creating noise for them)

**Reply by mnalis** [reply]
- want to use short (and thus) cryptic notes like "S15 H+ R2" instead of typing longer text (and taking photo is not wanted as mobile might be locked to bicycle handlebar or to avoid provoking questions)

**Reply by mnalis** [reply]
- you don't speak English nor the language of the country in question, but noticed problem that you want to solve later when you get home

**Reply by mnalis** [reply]
- StreetComplete (sometimes) does not allow you to see public notes that you created yourself, and user wants to

**Reply by rhhsm** [reply]
- you don't want to bother other mappers with your notes because you anticipate they will start complaining about them

**Reply by Strubbl** [reply]
a high number of notes in a small area is possible this way. 

Othwise a high number of notes in a small area and short time bothers other mappers because 
* they receive a notification for every created note in their watched area (e.g. via RSS) or 
* they see the forest of notes as pollution of the osm.org map or
* they misinterpret the open note as a task i am assigning for them although i am a mapper and can solve this by myself

that's what i was told when creating ~20 notes in a small area.

**Comment by rhhsm** [comment]
This discussion has been quiet for some time, maybe because we're almost there with the v32 option to list and revert edits and to choose manual upload. In this way, notes which are meant to be private can be solved and then reverted before the other edits from the survey trip are uploaded. This works fine for me, except that it's cumbersome to find the notes in a list that can become very long (several hundred items). I've be fully satisfied if notes could be made more easily visible by the author: now notes are "notes to everybody except self". Could notes be made visible to the author who created them, and a simple revert/delete button added to them, only visible to the author, so they can be easily deleted before upload?

**Reply by mnalis** [reply]
Also related: GPX notes in @Helium314 fork mentioned at https://github.com/streetcomplete/StreetComplete/discussions/3003#discussioncomment-963592 are a variant of private notes?

**Reply by rhhsm** [reply]
Another weakness of the present possibility to search for "notes to self" in the list of edits is that you can only see the text of the note to be reverted, but not any photos you may have taken as illustration. At present, images can only be seen by the note's author in a Windows editor, after upload.

**Comment by ppete2** [comment]
The last weeks I noticed a few new OSM-mappers which are using StreetComplete and opened DOZENS of public notes each day. And to be clear: This aren't notes as they should be used: to report an error in the data and ask OTHER mappers for help or their opinion. No - they have been just used as private reminders, to fix some SC-question at home. This is not what OSM-Notes are for: "Dont' use them just as private reminders": https://wiki.openstreetmap.org/wiki/DE:Notes  

This behaviour of StreetComplete is really dissappointing, cause the map is cluttered by such private notes, and complicates locating the "real" notes adressing other mappers.

I already wrote to such newbies, and in general they answer: They didn't know about this fact and would be happy if a "reminder" function within SC would exist. 

Please introduce this feature "private notes", to make SC users being able to create such reminder-notes which are just stored on their local phone.

**Reply by mnalis** [reply]
> The last weeks I noticed a few new OSM-mappers which are using StreetComplete and opened DOZENS of public notes each day.

@ppete2 could you link to few of them, please?

> This aren't notes as they should be used: to report an error in the data and ask OTHER mappers for help or their opinion.

I'm not sure about `DE` translation; but default English version of that page says _"Although you can use notes as a reminder to yourself, you are also inviting others to look at it. Descriptions must make sense to other people."_ which seems somewhat different definition. But I absolutely agree with you that public Notes should be solvable by anyone, not just their creator.

> This behaviour of StreetComplete is really dissappointing

You mean, behavior of StreetComplete _users_? StreetComplete app itself allows creating OSM Notes, as do many other [OSM editing tools](https://wiki.openstreetmap.org/wiki/Editors). Especially, when creating a new note, StreetComplete tells the user: _"You can leave a note to let other mappers know about an issue here"_. 
I don't know if German translation is lacking there, perhaps that can be improved? Or do you have suggestion what it should say to be better understandable, or how to make it more visible?

> Please introduce this feature "private notes", to make SC users being able to create such reminder-notes which are just stored on their local phone.

That feature is implemented in [SCEE](https://wiki.openstreetmap.org/wiki/SCEE) fork of StreetComplete, if someone is interested in using that functionality right now.

(manual ref.: @ppete2 complaint is also crossposted at https://github.com/streetcomplete/StreetComplete/issues/1992#issuecomment-1747096909; as GitHub does not implement linking issues with discussion I'm also mentioning it here)

**Reply by mnalis** [reply]
Also, as a workaround, if one does not find Notes left by StreetComplete (or EveryDoor or any other tools that leaves a signature/hashtag in Notes) of high enough quality and wishes to ignore them (and/or only ignore Notes from specific users etc), one can use various OSM tools to filter them out, like:

- https://antonkhorev.github.io/osm-note-viewer/ for one-time review of filtered notes (because [NotesReview](https://ent8r.github.io/NotesReview/), which is easier to use, does not yet allow for inverted filtering)
- for regular area patrolling, I'd recommend using excellent https://tyrasd.github.io/osm-qa-feeds/ which can produce [RSS](https://en.wikipedia.org/wiki/RSS) feed of new notes for specific area (as well as many other QualityAssurance tools for areas!), which has many benefits (there are many RSS readers to choose and they allow for custom filtering, but most importantly, you only need to check new notes _once_ instead of re-reading them every day/week, and you can favourite or mark them as read/unread as you wish. **Huge time saver**, and makes many things much easier like marking some notes for later review etc!)

**Reply by Helium314** [reply]
I agree here, using proper tools for note management is a far better solution than pestering users about not creating notes with wrongly claiming the are "private notes".
Also has the advantage that users keep adding details via their notes, whether the solve the notes themselves or not.

**Comment by margin-auto** [comment]
I'd strongly support this feature. I often use my own abbreviations in notes and don't think, they're useful for other users. I also use Osmand for taking notes, but .... well ... it's Osmand with its own UI and userfriendlieness....

**Reply by mnalis** [reply]
Thanks for your feedback, @margin-auto. 

Note however that this GitHub discussion is formatted in several different sections, each of them concerned with a particular aspect of the feature (e.g. _'How to create a "private note"?'_, _'How to find private notes created?'_ etc.) to inform devs what would work for users, and what wouldn't.

So intended usage for feedback is that you would go through each section, read it, and add thumb-up ( :+1: ) on a implementation that would work for you for _each_ of those sections (or add another option under that section if none of the provided would be good for you).

---

That being said, I would advise all parties interested in such "private notes" / "notes to self" to install [SCEE](https://wiki.openstreetmap.org/wiki/SCEE) (StreetComplete _"Expert edition"_ fork), go into _"Settings"_ and in _"SCEE settings"_ section _"Notes"_ enable _"Show GPX note button"_. 

Do **not be afraid** of the _"Expert"_ in the name, if you enable only that setting, the app will look and behave 99% like regular StreetComplete - just when creating a new note you'll have an extra button for _"GPX (private) note"_. So feel free to try it out, and _then_ base your :+1: feedback on the sections above based on your experience in SCEE. You can also have both StreetComplete and SCEE installed at the same time, without them interfering with each other.

Thanks for your effort trying to make StreetComplete better!
